### PR TITLE
[BEAM-3981] Futurize coders subpackage

### PR DESCRIPTION
**Pull request by RobbeSneyders**
This pull request prepares the coders subpackage for Python 3 support. This pull request is the first of a series in which all subpackages will be updated using the same approach.
This approach has been documented [here](https://docs.google.com/document/d/1xDG0MWVlDKDPu_IW9gtMvxi2S9I0GB0VDTkPhjXT0nE/edit?usp=drive_web&ouid=104989467967390032603) and the WIP pull request can be found at #4990.

The used approach can be summarized as follows:

- The future package provides tools to forward-port our code, while the six package focuses more on backporting. Whenever possible, we will therefore use the future package instead of the six package.
The future package provides backported Python 3 builtins, which can be used to write Python 3 style code with python 2 compatibility.

- One of the biggest problems in porting python 2 code to python 3, is the changed handling of strings and bytes.

  - The future package provides backported Python 3 `str` and `bytes` types. While these can be convenient to write Python 2/3 compatible code in end products, we don’t believe they are the right choice for Beam.
These backported types are new classes, subclassed from existing Python 2 builtins (e.g. `from future.builtins import int` imports the class `future.builtins.newint.newint`, which is a subclass op the Python 2 `long` type).
While these new classes behave like the Python 3 types, they don’t give the same results when used in type checks, which are constantly used in beam (e.g. typecoders, typechecks, …)

  - Instead, we propose to rewrite everything using the default `str` and `bytes` types. On Python 2, the `bytes` type is an alias for `str`. On Python 3, `str` is equivalent to Python 2 `unicode`.
A consistent behaviour between Python 2/3 can be reached by using the `bytes` type whenever `str` behaviour is desired on Python 2 and `bytes` behaviour is desired on Python 3 (= bytes data).
The `unicode` type can be used whenever `unicode` behaviour is desired on Python 2 and `str` behaviour is desired on Python 3 (= text data). The `unicode` type is not available in Python 3, which can be solved by adding
    ```
    Try:
      unicode           # pylint: disable=unicode-builtin
    except NameError:
      unicode = str
    ```
    at the top of the module.

  - All string literals which represent `bytes`, should be marked as `b’’`. String literals representing `unicode` in test modules should not be marked u’’. These will automatically be interpreted as `unicode` literals in Python 3, but we still want to test for unmarked Python 2 code.
Do not use the `from __future__ import unicode_literals` import since [its changes are too implicit and introduces a risk of subtle regressions on python 2](http://python-future.org/unicode_literals.html). 

- The used approach for the long / int types is equivalent to the `unicode` / `str` approach outlined above.
The `long` type is not available in Python 3, since `int` now has `long` behaviour. This can be solved by adding
    ```
    Try:
      long          # pylint: disable=long-builtin
    except NameError:
      long = int
    ```
    at the top of the module.

- Regression should be avoided as much as possible between the application of step 2 and step 3. This document proposes to take following measures to keep the probability of regression as low as possible:
  - Add the following import to the top of every module:
`from __future__ import absolute_import`
We can also add following imports to the top of every measure. This will ensures that no new code can be added using for instance the old python 2 division and adds consistency across modules. We would like to hear the community’s opinion on this.
    ```
    from __future__ import division
    from __future__ import print_function
    ```
  - A new tox environment has been added which runs pylint --py3k to check for python 3 compatibility.

- `range()` and `iteritems()` were not imported in from future.builtins in coder_impl.py to avoid performance regression in Cython compiled code.

@aaltay @charlesccychen

### CODE REVIEW

**Comment by angoenka** [comment]
R: @robertwb

**Comment by aaltay** [comment]
R: @charlesccychen  cc: @tvalentyn 

Some high level comments:
- six could be handy in some cases, for example `six.string_types` could be used instead of
```
Try:
  unicode           # pylint: disable=unicode-builtin
except NameError:
  unicode = str
```
What do you think about that?
- Is there a reason to change cython version?

**Comment by RobbeSneyders** [comment]
- I tried to remove all six references so we don't have 2 dependencies for py2/3 compatibility. If the six method is preferred, I can revert the changes.

- The previous cython version is not compatible with future.builtin types. Without this bump, any pure python cythonized file cannot use the builtins.

**Comment by RobbeSneyders** [comment]
Working on some other subpackages, I've come across an additional reason to upgrade the cython version. 

Currently, the `__cmp__` method is used in cythonized python files instead of the `__eq__` method because cython did not support the `__eq__` or any other special comparison methods. In Python 3, support for `__cmp__` is gone, so this workaround will not work anymore.

[Cython has added support for the special comparison methods in version 0.27.0](http://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#rich-comparisons), which allows us to use `__eq__` again.

File: sdks/python/apache_beam/coders/coders.py [hunk 1]
```diff
@@ -216,6 +222,9 @@ def __eq__(self, other):
             and self._dict_without_impl() == other._dict_without_impl())
     # pylint: enable=protected-access
 
+  def __hash__(self):
+    return hash(type(self))
```
Comment (charlesccychen) [inline sdks/python/apache_beam/coders/coders.py]: Any particular reason for this change?  Previously, the hash would default to `object.__hash__`, which tries to give a different hash code for each instance.  This change would give the same hash code for each class, which may not be desirable, since coders in general could be parameterized.

File: sdks/python/apache_beam/coders/coders.py [hunk 2]
```diff
@@ -309,11 +318,6 @@ class ToStringCoder(Coder):
   """A default string coder used if no sink coder is specified."""
 
   def encode(self, value):
-    try:               # Python 2
-      if isinstance(value, unicode):
-        return value.encode('utf-8')
```
Comment (charlesccychen) [inline sdks/python/apache_beam/coders/coders.py]: Should we do the same `unicode = str` thing here?  (In the new version, we will raise an error if the value is a non-ascii unicode string; eg: `str(u'😋')` raises an error, while this worked before.)

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 3]
```diff
@@ -279,21 +282,21 @@ def get_estimated_size_and_observables(self, value, nested=False):
...
-      stream.write_byte(STR_TYPE)
+    elif t is bytes:
+      stream.write_byte(BYTES_TYPE)
       stream.write(value, nested)
-    elif t is six.text_type:
-      unicode_value = value  # for typing
-      stream.write_byte(UNICODE_TYPE)
-      stream.write(unicode_value.encode('utf-8'), nested)
+    elif t is unicode:
+      text_value = value  # for typing
```
Comment (charlesccychen) [inline sdks/python/apache_beam/coders/coder_impl.py]: Can you use the same `try: unicode except: unicode = str` in the corresponding part of the .pxd file so that the type annotation directive for `text_value` is respected?

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 4]
```diff
@@ -304,7 +307,7 @@ def encode_to_stream(self, value, stream, nested):
       dict_value = value  # for typing
       stream.write_byte(DICT_TYPE)
       stream.write_var_int64(len(dict_value))
-      for k, v in dict_value.iteritems():
+      for k, v in dict_value.items():
```
Comment (charlesccychen) [inline sdks/python/apache_beam/coders/coder_impl.py]: Any particular reason for the `iteritems()` -> `items()` change?

Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coder_impl.py]: `iteritems()` was replaced by `items()` in Python 3 dict

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 3 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coder_impl.py]: I tried this, but it throws a 'not a type' error for `unicode`.

File: sdks/python/apache_beam/coders/coders.py (hunk 2 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coders.py]: The `unicode = str` trick won't work here on Python 3:
```
>>> str('test'.encode('utf-8'))
"b'test'"
```
I've reverted this change to the previous solution

File: sdks/python/apache_beam/coders/coders.py (hunk 1 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coders.py]: You're right. I've changed the hash to match the `__eq__` check.

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 4 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coder_impl.py]: `iteritems()` isn't available anymore in Python 3. Instead, `items()` returns an iterator. I've replaced it with `iteritems(dict)` from future.utils, which returns an iterator on both versions.

File: sdks/python/apache_beam/coders/coders.py [hunk 5]
```diff
@@ -210,11 +216,15 @@ def as_cloud_object(self):
...
     return self.__class__.__name__
 
+  # pylint: disable=protected-access
   def __eq__(self, other):
-    # pylint: disable=protected-access
     return (self.__class__ == other.__class__
             and self._dict_without_impl() == other._dict_without_impl())
-    # pylint: enable=protected-access
+
+  def __hash__(self):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/coders/coders.py]: Is this change required by Python3 migration or we are just fixing an omission that __hash__ was not previously defined, while __eq__ was?

File: sdks/python/apache_beam/coders/coders_test_common.py [hunk 6]
```diff
@@ -103,7 +104,7 @@ def test_custom_coder(self):
 
     self.check_coder(CustomCoder(), 1, -10, 5)
     self.check_coder(coders.TupleCoder((CustomCoder(), coders.BytesCoder())),
-                     (1, 'a'), (-10, 'b'), (5, 'c'))
+                     (1, b'a'), (-10, b'b'), (5, b'c'))
```
Comment (tvalentyn) [inline sdks/python/apache_beam/coders/coders_test_common.py]: Probably not critical, but looks like 'a' is not replaced with b'a' here - are these changes done by some tool or manually?

File: sdks/python/apache_beam/coders/coders_test_common.py [hunk 7]
```diff
@@ -129,7 +130,7 @@ def test_dill_coder(self):
 
   def test_fast_primitives_coder(self):
     coder = coders.FastPrimitivesCoder(coders.SingletonCoder(len))
-    self.check_coder(coder, None, 1, -1, 1.5, 'str\0str', u'unicode\0\u0101')
+    self.check_coder(coder, None, 1, -1, 1.5, b'str\0str', u'unicode\0\u0101')
     self.check_coder(coder, (), (1, 2, 3))
     self.check_coder(coder, [], [1, 2, 3])
     self.check_coder(coder, dict(), {'a': 'b'}, {0: dict(), 1: len})
```
Comment (tvalentyn) [inline sdks/python/apache_beam/coders/coders_test_common.py]: Also here 'a' and 'b' are not bytestrings.

File: sdks/python/tox.ini [hunk 8]
```diff
@@ -88,6 +89,21 @@ commands =
   pip --version
   time {toxinidir}/run_pylint.sh
 
+[testenv:py27-lint3]
```
Comment (tvalentyn) [inline sdks/python/tox.ini]: Can we add a comment how this is different from py3-lint? Or perhaps we don't need both of them?

File: sdks/python/apache_beam/coders/coders.py (hunk 5 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coders.py]: On Python 2, `__hash__` defaults to give a different value for each instance. On Python 3, `__hash__` defaults to `None` if `__eq__` is implemented. By implementing `__hash__`, we get consistent behavior on both versions.

File: sdks/python/tox.ini (hunk 8 above)
Comment (RobbeSneyders) [reply sdks/python/tox.ini]: py27-lint3 checks for portability issues, while py3-lint checks for python 3 issues. I'll add a comment to clarify.

File: sdks/python/apache_beam/coders/coders_test_common.py (hunk 6 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coders_test_common.py]: I guess this is aimed at the 'a' in line 109? 
The marking of the strings as bytes literals is done manually. I've only marked strings as bytes literals when it's clear that they're meant to represent bytes (when testing BytesCoder, when the content of the string are clearly bytes, ...).  When a string is not marked, it represents `str` on both versions, which seems ok for the 'a' at line 109 for example.

File: sdks/python/apache_beam/coders/coders_test_common.py (hunk 7 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coders_test_common.py]: See comment above.
The unmarked 'a' and 'b' here represent `str` on both versions, which seems ok for this test.

**Comment by RobbeSneyders** [comment]
Thanks for the reviews @charlesccychen, @tvalentyn. I've tried to address all of your comments and have committed some changes based on your input.

**Comment by RobbeSneyders** [comment]
run java precommit

File: sdks/python/run_pylint.sh [hunk 9]
```diff
@@ -118,4 +99,4 @@ if [ -n "${TESTS_MISSING_MAIN}" ]; then
   done
   echo
   exit 1
-fi
+fi
```
Comment (tvalentyn) [inline sdks/python/run_pylint.sh]: Let's add a new line at the end of file here and in sdks/python/run_pylint_2to3.sh.

Comment (RobbeSneyders) [reply sdks/python/run_pylint.sh]: done

**Comment by tvalentyn** [comment]
Thank you Robbe! The PR and approach look good to me.

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by aaltay** [comment]
I see the following error in the output

```
I      /tmp/pip-cVZPmP-build/setup.py:79: UserWarning: You are using version 0.27.3 of cython. However, version 0.28.1 is recommended. 
I        _CYTHON_VERSION, REQUIRED_CYTHON_VERSION 
I       
I      Error compiling Cython file: 
I      ------------------------------------------------------------ 
I      ... 
I      except NameError:   # Python 3 
I        long = int 
I        unicode = str 
I       
I       
I      class CoderImpl(object): 
I      ^ 
I      ------------------------------------------------------------ 
I       
I      apache_beam/coders/coder_impl.py:68:0: 'object' is not a type name 
I      Compiling apache_beam/coders/stream.pyx because it changed. 
I      Compiling apache_beam/runners/worker/statesampler_fast.pyx because it changed. 
I      Compiling apache_beam/coders/coder_impl.py because it changed. 
I      Compiling apache_beam/metrics/execution.py because it changed. 
I      Compiling apache_beam/runners/common.py because it changed. 
I      Compiling apache_beam/runners/worker/logger.py because it changed. 
I      Compiling apache_beam/runners/worker/opcounters.py because it changed. 
I      Compiling apache_beam/runners/worker/operations.py because it changed. 
I      Compiling apache_beam/transforms/cy_combiners.py because it changed. 
I      Compiling apache_beam/utils/counters.py because it changed. 
I      Compiling apache_beam/utils/windowed_value.py because it changed. 
I      [ 1/11] Cythonizing apache_beam/coders/coder_impl.py 
I      Traceback (most recent call last): 
I        File "<string>", line 1, in <module> 
I        File "/tmp/pip-cVZPmP-build/setup.py", line 204, in <module> 
I          'apache_beam/utils/windowed_value.py', 
I        File "/usr/local/lib/python2.7/dist-packages/Cython/Build/Dependencies.py", line 1039, in cythonize 
I          cythonize_one(*args) 
I        File "/usr/local/lib/python2.7/dist-packages/Cython/Build/Dependencies.py", line 1161, in cythonize_one 
I          raise CompileError(None, pyx_file) 
I      Cython.Compiler.Errors.CompileError: apache_beam/coders/coder_impl.py 
I       
I      ---------------------------------------- 
```

I assume this is because The new code requires the new Cython version however dataflow workers do not have it.

@tvalentyn Could you upgrade the workers at head to use the 0.28.1 cython version?

**Comment by tvalentyn** [comment]
Containers are released and PR #5131 is out to upgrade them on master, we can also apply the changes from that PR here as well and rerun the postcommit suite.

**Comment by aaltay** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
With  #5131 merged, we probably need to rebase this PR off the current masters, for the ValidatesRunner tests to pass. @RobbeSneyders can we do that please?

**Comment by tvalentyn** [comment]
Actually, looking at Jenkins logs I see that Jenkins already merges this PR with latest commit on master when we run a Postcommit suite:  
> Checking out Revision 1b8df077a60fc0188d786a146d5c5edb9eb2732f (refs/remotes/origin/pr/5053/merge)
>  > git config core.sparsecheckout # timeout=10
>  > git checkout -f 1b8df077a60fc0188d786a146d5c5edb9eb2732f
> Commit message: "Merge 6909ff3d2b1802c0712d8236692c0c7110d9b250 into e1c526d7f88add0aa44635a56223a4907b81e86b"
> ...
>

**Comment by tvalentyn** [comment]
There is a proto generation error in the ValidatesRunner tests, which seems to be unrelated to this PR, I also see that error in other Postcommit  test suites.

**Comment by RobbeSneyders** [comment]
I've rebased the branch anyway. Is there anything else I should do before we can merge? Like squash some commits?

**Comment by tvalentyn** [comment]
Yes, please squash the commits; I am looking into postcommit issue, hopefully that does not affect this PR.
Thank you.

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by RobbeSneyders** [comment]
Run Python Dataflow ValidatesRunner

**Comment by RobbeSneyders** [comment]
Run Python Precommit

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
@tvalentyn if the test issues are unrelated to this PR and related to the current ongoing Jenkins issues should we move forward with it? (I suggest that we run a few manual tests locally to check whether this PR introduces issues or not.)

**Comment by tvalentyn** [comment]
https://github.com/apache/beam/pull/5163 should fix Jenkins build, let's rerun the tests again once that is merged. We can continue working on other Python-3 changes in parallel, I don't anticipate changes to this PR at this point but would like to get a signal from ValidatesRunner suites. Happy to test manually if Jenkins build does not get resolve within next day.

**Comment by tvalentyn** [comment]
Alright, Mr. Jenkins, be kind and patient to us: Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Looks like this is ready to merge, @aaltay !

**Comment by RobbeSneyders** [comment]
ping @aaltay

**Comment by tvalentyn** [comment]
Cython issue has a workaround and is no longer a blocker. 

I observed a 15% performance degradation on Dataflow runner with this PR & Cython upgrade, however this was a one-off benchmark run and the data is inconclusive. I am starting multiple runs to confirm or rule out the regression here.

**Comment by tvalentyn** [comment]
@RobbeSneyders I should have an update on performance later today. We'd like to make sure we understand if there are any performance implication since this code is on the data path. 
I am hoping that the merge is not blocking you from other py3 efforts, but let me know if this is a hard dependency for you at the moment. Thank you.

**Comment by RobbeSneyders** [comment]
Ok, thank you.

PRs of other packages are blocked by this one because of the added lint tests and the cython upgrade. I have already ported 3 extra packages which are waiting to be submitted, but I have paused extra porting work so I won't have to redo too much work if anything changes.
When this PR is merged, I will reach out to the mailing list to work on other packages in parallel so we can move forward quickly.

**Comment by tvalentyn** [comment]
After collecting additional datapoints on the same benchmark, the regression seems to be on the order of 7%, I am doing more A/B testing to confirm the exact changes that contribute to it.

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 10]
```diff
@@ -253,8 +258,8 @@ def decode(self, encoded):
 NONE_TYPE = 0
 INT_TYPE = 1
 FLOAT_TYPE = 2
-STR_TYPE = 3
-UNICODE_TYPE = 4
+BYTES_TYPE = 3
+TEXT_TYPE = 4
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coder_impl.py]: I would leave this as UNICODE_TYPE. TEXT_TYPE is ambiguous wrt. Python 2 vs. 3.

Comment (robertwb) [reply sdks/python/apache_beam/coders/coder_impl.py]: Also, any renaming here should be reflected in the .pxd file. That may account for some of the perf issues. 

Running "python -m cython -a apache_beam/coders/coder_impl.py" could be useful for debugging this too.

**Comment by tvalentyn** [comment]
I have tried to revert several statements  that might trigger a regression (hash functions, dropped typing annotation in cython code, explicit encode() statements, and recent commit). I also started a suite with more wide reverts. I also plan to take a look at generated cython code by robertwb's suggestion. Current suite I am running takes around 8hours to complete, which reduces feedback look significantly. If I still don't see the culprit tomorrow, I might try to reproduce the regression in a microbenchmark that specifically targets the code we are modifying code to reduce the feedback loop. There's not that many things left to revert.

**Comment by RobbeSneyders** [comment]
Ok thank you @tvalentyn.

A small performance regression in Python 2 can be expected with the chosen approach. With the use of the future package, we focus more on writing native Python 3 and backporting this to Python 2. I believe this is the best choice since Python 3 will provide a performance improvement and support for Python 2 will be dropped in less than 2 years. 

I have applied the changes proposed by @robertwb, which will hopefully reduce the amount of regression.

**Comment by tvalentyn** [comment]
Based on the information I currently have, my best guess is that regression is caused by changes in Cython's interpretation of Py3 implementations of range() and iteritems() in coder_impl.py. We can see a difference in code generated by Cython here:

https://docs.google.com/document/d/1S-oeqJGiMHt_L3iCgr9dYfQdR0_ukQE25mcvK-BqudU/edit#heading=h.xkt18qn9ofn1

I am seeing 7% regression in the median on the benchmarks I've been running, and I was able to see the regression go away with these changes reverted. 

We can also try to benchmark these methods in isolation to confirm and rule out the variance of test setup. I might not have time to do it this week.

@robertwb , do you  think anything can be done to improve Cython translation here?

**Comment by RobbeSneyders** [comment]
I think we can remove the `range()` import from coders_impl.py without changing any behavior. We will have to disable the lint check for this file though.

An alternative to the iteritems() import is using `for k, v in dict_value.items()`. This will have improved performance in Python 3, but will return a list on Python 2. This might however still perform better with Cython compared to the `iteritems()` import. Could you check this, @tvalentyn?

**Comment by tvalentyn** [comment]
@RobbeSneyders I think it's fine to make one-off exceptions for coders_impl.py, and annotating them with inline exceptions from lint checker for the two changes we discussed. We don't expect these kind of exceptions for most other parts of codebase, but wanted to do our due diligence on this package since it's on critical path.

I agree with you, let's remove the `range()` import. For `iteritems`  how about we do something along the lines: 

```
if (...python 2...):
  items = dict_value.iteritems()   # no-py3lint: allow iteritems()
else:
  items = dict_value.items()
for k,v in items:
...
```
I think with these changes we can proceed with the merge, I will do benchmarking in parallel and we can clean this up later.

**Comment by RobbeSneyders** [comment]
Ok, great. I have pushed the new changes.
Seems like the javaPreCommit failing is unrelated.

**Comment by tvalentyn** [comment]
Thanks Robbe, the change looks good to me. I noticed we did not whitelist coders_impl.py from linter given that import of range was missing(). So, linter does not require this import?

**Comment by tvalentyn** [comment]
Can you please add a comment in the file and/or PR description that we intentionally did not import range and use iteritems() suspecting possible performance regression in compiled Cython code?

**Comment by RobbeSneyders** [comment]
I have added comments in the file and added a remark in the documentation of the used approach.

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
@RobbeSneyders Thank you for the comments.

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesRunner

**Comment by tvalentyn** [comment]
ValidatesRunner suite just takes time to start. It failed due to an unrelated issue, filed: https://issues.apache.org/jira/browse/INFRA-16508 to clarify. I'll do a sanity check locally.

**Comment by robertwb** [comment]
Regarding items vs. iteritems, let's just go with dict_value.items everywhere. There's little if any need for explicit iteritems in this case (and if dict is typed, I think Cython optimizes `for k, v in dict_value.items()` optimally without the intermediate list anyways). 

Though macrobenchmarks can be good for final validation, this is the kind of code that could probably benefit from some pretty simple microbenchmarks.

File: sdks/python/apache_beam/coders/coders.py (hunk 5 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/coders/coders.py]: Revisiting this in light of other PRs. I think, it would be safer to guarantee the contract that hash does not change for the same object if we compute it here based on object type, sent https://github.com/apache/beam/pull/5390. 
Another possibility to guarantee consistent behavior between Python 2 and 3 would be to set `__hash__ = None` if we can infer that a class is obviously non-hashable.

Comment (RobbeSneyders) [reply sdks/python/apache_beam/coders/coders.py]: We can also explicitly implement the default python 2 hash, which is guaranteed to stay the same for an object:
`hash(id(self))`

Comment (tvalentyn) [reply sdks/python/apache_beam/coders/coders.py]: That's true. Although that wouldn't honor the contract between __eq__ and __hash__.

**Comment by tvalentyn** [comment]
An update on `dict.iteritems` vs `dict.items()` vs `future.utils.iteritems(dict)` - I did more performance testing of encode-decode operation using a microbenchmark (currently in flight: https://github.com/apache/beam/pull/5565). 

I don't observe a difference in performance of `dict.iteritems()` and `future.utils.iteritems(dict)`. 

As far as `dict.items()` vs `dict.iteritems()` goes, I saw a 2x performance slowdown in coder implementation with dict.items() for dictionaries with over 100000 entries, but did not observe a significant difference on dictionaries with 10000 entries or less. That said I think it would not hurt to keep using `iteritems()` for Python 2 as we do now.

With `future.utils.iteritems()`:

```
Median time cost:
Dict[int, int], FastPrimitiveCoder         : per element median time cost: 3.27529e-07 sec                                                  
```
With `dict.iteritems()`:
```
Median time cost:
Dict[int, int], FastPrimitiveCoder         : per element median time cost: 3.4485e-07 sec
```
With `dict.items()`:

```
Median time cost:
Dict[int, int], FastPrimitiveCoder         : per element median time cost: 7.3393e-07 sec
```

I also observe a 2.5x degradation in coder implementation with `builtins.range()` compared to `range()` on lists as small as 1000 - 10000 elements. I did not try smaller lists.

With 10000 elements, python 2 `range()`:
```
Median time cost:
List[int], FastPrimitiveCoder              : per element median time cost: 1.17695e-07 sec
```

With `builtins.range():`
```
Median time cost:
List[int], FastPrimitiveCoder              : per element median time cost: 3.22402e-07 sec
```

We should try to use microbenchmarks for performance evaluations moving forward since they can provide feedback in a matter of seconds.

**Comment by robertwb** [comment]
`dict.items()` should have the same performance when compiled with Cython, as it unpacks it into item-iterating code without the function call (or intermediate object generation). Having two cases for Py2 and Py3 will break this optimization (as well as the fact that the case checking code could be (relatively) expensive for small dicts.)

**Comment by tvalentyn** [comment]
Hmm. I am pretty sure my microbenchmark uses a Cython codepath since in order for any code change to take effect I have to run  `python setup.py build_ext --inplace` to recompile associated C extensions.  I checked once again and I do see a 2x slowdown with `items()` once the size of the dictionary goes more than 100000 elements. Here's code generated by Cython: a https://docs.google.com/document/d/1S-oeqJGiMHt_L3iCgr9dYfQdR0_ukQE25mcvK-BqudU/edit#heading=h.drcukhvo4hd6. Perhaps the slowdown is related to materializing the list? 

My microbenchmark setup is here: https://github.com/apache/beam/compare/master...tvalentyn:coders_dict_microbencmark?expand=1

**Comment by tvalentyn** [comment]
For the record, https://github.com/apache/beam/pull/5586 improves the performance of `dict.items()` with a Cython directive to use Python 3 interpretation.
